### PR TITLE
feat(internal): use git to determine branch modification status

### LINKED ISSUE #6693: Remove pr.isModified and implement isBranchModified() in git

### PR DESCRIPTION
**Pull request by rarkins**
Uses git log to determine if the Renovate’s gitAuthor was the author of the most recent commit per branch. If it is, then the branch/PR is considered unmodified.

Closes #6693

BREAKING CHANGE: If Renovate shares a git account with another real or bot user which commits to Renovate’s branches, Renovate may no longer detect that as a modified branch. It is recommended to use an account that is unique to Renovate, or to configure a distinct `gitAuthor` (e.g. `bot@renovateapp.com`).

### CODE REVIEW

**Comment by rarkins** [comment]
Another PR which removes a lot of duplicated/unnecessary logic from each platform. However there's a small chance that some users were relying on Renovate's previously more complicated check for modified branches, e.g. if they ran Renovate using a real user's account, and that real user might occasionally commit over the top of Renovate, or if they had some other bot that adds to Renovate's commits using the same account. Both of these can be worked around by simply configuring a distinct `gitAuthor`.

**Comment by rarkins** [comment]
@JamieMagee @viceice keeping this as Draft so we don't merge it by accident, but it's ready for you to take a look.

**Comment by rarkins** [comment]
One other downside is that Renovate will no longer ignore when people clicked the "Update branch" on GitHub. Currently Renovate tries to detect if merge commits were triggered by the UI and treat them as non-modified. However, considering the simplicity saving in our platform code and also that people on GitHub have easy access to "click to rebase", I think it's a worthy trade-off. Plus the old detection was a bit hacky anyway.

**Comment by viceice** [comment]
Ok,I need a little bit more time to review this tomorrow.

File: lib/util/git/index.ts [hunk 1]
```diff
@@ -395,6 +398,34 @@ export async function isBranchStale(branchName: string): Promise<boolean> {
...
+  if (config.branchIsModified[branchName] !== undefined) {
+    return config.branchIsModified[branchName];
+  }
+  if (!(await branchExists(branchName))) {
+    throw Error(
+      'Cannot check modification for branch that does not exist: ' + branchName
+    );
+  }
+  // Retrieve the author of the most recent commit
+  const lastAuthor = (
```
Comment (viceice) [inline lib/util/git/index.ts:412]: Can we also count the commits? Assume a user reordered his commit before the renovate commit.

File: lib/util/git/index.ts [hunk 2]
```diff
@@ -395,6 +398,34 @@ export async function isBranchStale(branchName: string): Promise<boolean> {
...
+  if (lastAuthor === gitAuthorEmail) {
+    // author matches - branch has not been modified
+    config.branchIsModified[branchName] = false;
+    return false;
+  }
+  logger.debug(
+    { branchName, lastAuthor, gitAuthorEmail },
+    'Last commit author does not match git author email - branch has been modified'
+  );
+  config.branchIsModified[branchName] = true;
```
Comment (viceice) [inline lib/util/git/index.ts:425]: Do we need to reset the cache after force rebase a modified PR?

File: lib/util/git/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/git/index.ts:412]: I'm not sure there's an easy way for us to do that at the git level. We might need to know what the target sha was in order to know when to stop counting commits.

File: lib/util/git/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/util/git/index.ts:425]: Value should be set to `false` after a commit.

File: lib/util/git/index.ts (hunk 1 above)
Comment (viceice) [reply lib/util/git/index.ts:412]: Possible

https://stackoverflow.com/a/11657647

```sh
git checkout master
git checkout -b test
<We do 3 commits>
git rev-list --count HEAD ^master
``|

Comment (viceice) [reply lib/util/git/index.ts:412]: `git rev-list --count develop..renovate/branch` should do it. Not yet tried

Comment (rarkins) [reply lib/util/git/index.ts:412]: It would be easy if it's up to date with master. I'm worried about the case where it branched off a few commits ago

Comment (rarkins) [reply lib/util/git/index.ts:412]: Maybe works:
```
❯ git rev-list --count fix/5951-retry-checkbox ^master
1
```

(the above branch was already merged and is out of date)

**Comment by JamieMagee** [comment]
Looks fine to me. Are you planning on making a major version bump when merging this?

**Review by JamieMagee — DISMISSED** [review]

**Comment by rarkins** [comment]
> Looks fine to me. Are you planning on making a major version bump when merging this?

I was planning to, because it could have a change of behavior for some. Although following @viceice's suggestion I was able to add back the "more than one commit means it's modified" logic so I think we're pretty close to parity now. The main difference is the "Update branch" detection on GitHub, which was always a bit of a hack.

**Review by viceice — COMMENTED** [review]
LGTM, needs deconflicting. @rarkins Did you real run againd some test repos?

**Review by viceice — DISMISSED** [review]
Anything missing here?

**Comment by rarkins** [comment]
A decision on whether breaking or not :D

I wouldn't mind changing the default git user too, a very minor "breaking" change. So maybe both at once?

**Comment by viceice** [comment]
I'm ok with a Major, so we're safe with a breaking.

**Comment by rarkins** [comment]
@viceice any chance these failures after merge are due to bugs in the new major version of simple-git?

File: lib/util/git/index.spec.ts [hunk 3]
```diff
@@ -36,6 +36,20 @@ describe('platform/git', () => {
     await repo.add(['future_file']);
     await repo.commit('future message');
 
+    await repo.checkoutBranch('renovate/modified_branch', 'master');
```
Comment (viceice) [inline lib/util/git/index.spec.ts:39]: This is broken and causes the test failures, see https://github.com/steveukx/git-js/issues/490.

We need 2.14.0 to fix that.

Comment (viceice) [reply lib/util/git/index.spec.ts:39]: https://github.com/steveukx/git-js/blob/master/CHANGELOG.md#2140---bug-fix-gitcheckoutbranch-fails-to-pass-commands-to-git-child-process

Comment (viceice) [reply lib/util/git/index.spec.ts:39]: #6818

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
merge together with got v11 ?

**Pull request merged**